local.go

```diff
@@ -338,130 +338,52 @@ func checkConvergence(loc *Location, iterType IterationType, stats *Stats, setti
-			return
-		}
-		if funcInfo.IsFunctionGradientHessian {
-			if loc.Hessian == nil {
-				loc.Hessian = mat64.NewSymDense(len(loc.X), nil)
-			}
-			loc.F = funcInfo.functionGradientHessian.FuncGradHess(loc.X, loc.Gradient, loc.Hessian)
-			stats.FuncGradHessEvaluations++
-			return
-		}
-	case HessEvaluation:
-		if funcInfo.IsHessian {
-			if different {
-				invalidate(loc, true, true, false)
-			}
-			funcInfo.hessian.Hess(loc.X, loc.Hessian)
-			stats.HessEvaluations++
-			return
-		}
-		if funcInfo.IsFunctionGradientHessian {
-			if loc.Gradient == nil {
-				loc.Gradient = make([]float64, len(loc.X))
-			}
-			loc.F = funcInfo.functionGradientHessian.FuncGradHess(loc.X, loc.Gradient, loc.Hessian)
-			stats.FuncGradHessEvaluations++
-			return
-		}
-	case FuncGradEvaluation:
-		if funcInfo.IsFunctionGradient {
-			if different {
-				invalidate(loc, false, false, true)
-			}
-			loc.F = funcInfo.functionGradient.FuncGrad(loc.X, loc.Gradient)
-			stats.FuncGradEvaluations++
-			return
-		}
-		if funcInfo.IsGradient {
-			if different {
-				invalidate(loc, false, false, true)
-			}
-			loc.F = funcInfo.function.Func(loc.X)
-			stats.FuncEvaluations++
-			funcInfo.gradient.Grad(loc.X, loc.Gradient)
-			stats.GradEvaluations++
-			return
-		}
-		if funcInfo.IsFunctionGradientHessian {
-			if loc.Hessian == nil {
-				loc.Hessian = mat64.NewSymDense(len(loc.X), nil)
-			}
-			loc.F = funcInfo.functionGradientHessian.FuncGradHess(loc.X, loc.Gradient, loc.Hessian)
-			stats.FuncGradHessEvaluations++
-			return
-		}
-	case FuncGradHessEvaluation:
-		if funcInfo.IsFunctionGradientHessian {
-			loc.F = funcInfo.functionGradientHessian.FuncGradHess(loc.X, loc.Gradient, loc.Hessian)
-			stats.FuncGradHessEvaluations++
-			return
-		}
-		if funcInfo.IsGradient && funcInfo.IsHessian {
-			loc.F = funcInfo.function.Func(loc.X)
-			stats.FuncEvaluations++
-			funcInfo.gradient.Grad(loc.X, loc.Gradient)
-			stats.GradEvaluations++
-			funcInfo.hessian.Hess(loc.X, loc.Hessian)
-			stats.HessEvaluations++
-			return
-		}
-	case NoEvaluation:
-		if different {
+func evaluate(f *functionInfo, evalType EvaluationType, xNext []float64, loc *Location, stats *Stats) {
+	if !floats.Equal(loc.X, xNext) {
+		if evalType == NoEvaluation {
 			// Optimizers should not request NoEvaluation at a new location.
 			// The intent and therefore an appropriate action are both unclear.
 			panic("optimize: no evaluation requested at new location")
 		}
-		return
-	default:
+		copy(loc.X, xNext)
+		invalidate(loc, evalType)
+	}
+
+	if evalType&FuncEvaluation != 0 {
+		loc.F = f.function.Func(loc.X)
+		stats.FuncEvaluations++
+		evalType &= ^FuncEvaluation
+	}
+	if evalType&GradEvaluation != 0 {
+		f.gradient.Grad(loc.X, loc.Gradient)
+		stats.GradEvaluations++
+		evalType &= ^GradEvaluation
+	}
+	if evalType&HessEvaluation != 0 {
+		f.hessian.Hess(loc.X, loc.Hessian)
+		stats.HessEvaluations++
+		evalType &= ^HessEvaluation
+	}
+
+	if evalType != NoEvaluation {
```

Comment:
I think this is a good way to check for bad evalTypes, but the error message should return the original evalType specified by the optimizer. Probably something like
````
toEval := evalType
if evalType&FuncEvaluation != 0 {
     toEval &= ^FuncEvaluation
} 
...
if toEval != NoEvaluation {
     panic(fmt.Sprintf("optimize: unknown evaluation type %v", evalType))   
}
````

Now that evalType is just a number, do we really need fmt?